### PR TITLE
Remove redundant ReadDir() calls in isSingleChild

### PR DESCRIPTION
**Pull request by lguychard**
Rel https://github.com/sourcegraph/customer/issues/23

A [customer](https://app.hubspot.com/contacts/2762526/company/554275594) reported that the files sidebar was slow to load, even though the "Files and directories" section of the main tree view, which displays basically the same information, was not.

The graphQL query for the sidebar (in `fetchTreeEntries()`) requested the following additional fields for each tree entry:

```
submodule {
    url
    commit
}
isSingleChild
```

In my tests, `submodule` did not affect performance, but `isSingleChild` did, causing up to a 50% perf hit for the same repository/commit/path even for small repositories.

Looking at the resolvers, I realised that the `IsSingleChild()` resolver performed a `git.ReadDir()` of the parent directory for each entry. This meant that the query was resulting in N+1 `git.ReadDir()` calls for the same directory, where N is the number of entries in that directory.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/git_tree_entry.go [hunk 1]
```diff
@@ -41,7 +41,8 @@ type GitTreeEntryResolver struct {
 	// the root, not the basename.
 	stat os.FileInfo
 
-	isRecursive bool // whether entries is populated recursively (otherwise just current level of hierarchy)
+	isRecursive   bool // whether entries is populated recursively (otherwise just current level of hierarchy)
+	isSingleChild *bool
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/git_tree_entry.go]:
```suggestion
	isSingleChild *bool // whether this is the single entry in its parent. Only set by the (&GitTreeEntryREsolver) entries.
```

**Comment by keegancsmith** [comment]
nice find

File: cmd/frontend/graphqlbackend/git_tree_entry.go [hunk 2]
```diff
@@ -208,6 +209,9 @@ func CreateFileInfo(path string, isDir bool) os.FileInfo {
 }
 
 func (r *GitTreeEntryResolver) IsSingleChild(ctx context.Context, args *gitTreeEntryConnectionArgs) (bool, error) {
+	if r.isSingleChild != nil {
+		return r.IsDirectory() && *r.isSingleChild, nil
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/git_tree_entry.go]: Super nitpicky, but technically you don't need to duplicate the `r.IsDirectory()` check here. You could just move the `if r.isSingleChild != nil {` check _after_ the `if !r.IsDirectory()`. But, again, not necessary and I also think something speaks to this version from a readability perspective.

**Comment by dadlerj** [comment]
@lguychard not a big deal at all, but just FYI, I changed your message to pull out the HubSpot URL from a markdown link. From our handbook:

>Do not use a custom markdown link for this. Link destinations are not searchable in GitHub issues—only the rendered text. Leaving the full URL visible in the text makes it possible to search a given company’s HubSpot ID to find all issues related to them.

https://about.sourcegraph.com/handbook/support/filing_customer_issues

If you ever see this in the future, please do the same!

**Comment by attfarhan** [comment]
Nice, thanks for the fix!

**Comment by tsenart** [comment]
Just came here to say: Welcome to the Core Services team @lguychard.

**Pull request merged**